### ISSUE TITLE
Failure to optimise `std::cmp::{min,max}` on `bool`

### ISSUE THREAD

**Issue by Kmeakin** [rust-lang/rust#114653]
I tried this code: ([godbolt](https://godbolt.org/z/vbdsMK1a5))

```rust
// DOES NOT optimise to `a & b`
pub fn min_std(a: bool, b: bool) -> bool {
    std::cmp::min(a, b)
}

// DOES NOT optimise to `a | b`
pub fn max_std(a: bool, b: bool) -> bool {
    std::cmp::max(a, b)
}

// DOES optimise to `a & b`
pub fn min_if(a: bool, b: bool) -> bool {
    if (a < b) {
        return a;
    }
    b
}

// DOES optimise to `a | b` on x86_64
// DOES NOT optimise to `a | b` on AArch64
pub fn max_if(a: bool, b: bool) -> bool {
    if (a > b) {
        return a;
    }
    b
}

pub fn min_and(a: bool, b: bool) -> bool {
    a & b
}
pub fn max_or(a: bool, b: bool) -> bool {
    a | b
}

```

I expected to see this happen:
All implementations of max should optimise to `a & b`. All implementations of min should optimise to `a | b`

Instead, this happened:
The generated assembly code is suboptimal for `min_std`, `max_std`, `max_ternary` and `max_if` (on AArch64)

`min_std` and `max_std` can be fixed by overriding the default implementation of `min` and `max` for `bool`:
```rust
impl Ord for bool {
    fn min(self, other: Self) -> Self { a & b }
    fn max(self, other: Self) -> Self { a | b }
}
```
But it might be worth investigating the LLVM-IR that rustc generates in `min_if`/`max_if` too to see why it isn't optimised to `llvm.umin()`/`llvm.umax()`

### Meta

`rustc --version --verbose`:
```
rustc 1.73.0-nightly (f88a8b71c 2023-08-08)
binary: rustc
commit-hash: f88a8b71cebb730cbd5058c45ebcae1d4d9be377
commit-date: 2023-08-08
host: x86_64-unknown-linux-gnu
release: 1.73.0-nightly
LLVM version: 17.0.0
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

**Comment by Kmeakin** [comment]
Corresponding LLVM issue: https://github.com/llvm/llvm-project/issues/64537

**Comment by Qi-Hu** [comment]
I made a PR to llvm-project to optimize the `max_if` case: https://github.com/llvm/llvm-project/pull/66394

**Comment by Qi-Hu** [comment]
> I made a PR to llvm-project to optimize the `max_if` case: [llvm/llvm-project#66394](https://github.com/llvm/llvm-project/pull/66394)

Actually this feature has already been implemented by [074f23e3e199](https://reviews.llvm.org/rG074f23e3e199a8067f0295302e07e9837c6f16d1). This issue can be closed.
